pkg: provide sock_ip support for lwip

miri64 opened the pull request:
This was taken out of #5802 and only provides the `sock_ip` part and the tests for it.

miri64 commented:
Needs adaptation to #5945.

kaspar030 commented:
Reassigning.

miri64 commented:
Adaption to https://github.com/RIOT-OS/RIOT/pull/6137#issuecomment-261506878 also needed...

miri64 commented:
Provided the same treatment to this PR as I did for #5937.

miri64 commented:
Ping?

miri64 commented:
This one isn't WIP anymore either.

PeterKietzmann commented on tests/lwip_sock_ip/README.md:
````diff
@@ -0,0 +1,33 @@
+To test both set the `LWIP_IPV4` and `LWIP_IPV6` to a non-zero value:
+
+```sh
+LWIP_IPV4=1 LWIP_IPV6 make all test
````
`LWIP_IPV6=1`

miri64 commented:
> @miri64 could you please adapt tests/lwip to use the introduced sock port? It is needed anyway and makes testing much easier. 

I plan to do this after these are merged. Since the different flavors of sock are divided up into multiple PRs this makes porting in all the different PRs much harder.

miri64 commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -0,0 +1,518 @@
+#error "lwip_sock needs IPv4 or IPv6 support"
+#endif
+
+static inline bool _af_not_supported(int af)
```
Will add some doc.

miri64 commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -0,0 +1,518 @@
+    if ((*out = netconn_new_with_proto_and_callback(type, proto, NULL)) == NULL) {
+        return -ENOMEM;
+    }
+#if SO_REUSE
```
This comes from lwIP. So I won't rename it.

PeterKietzmann commented:
@miri64 that makes totally sense! However, I assume you've already tested the implementation with a networking application, right? Most likely the application code will look like described [here](https://riot-os.org/api/group__net__sock__ip.html). But can you share the Makefile here (and in the other sock PRs)? That would make testing a lot easier.

miri64 commented:
> @miri64 that makes totally sense! However, I assume you've already tested the implementation with a networking application, right? Most likely the application code will look like described here. But can you share the Makefile here (and in the other sock PRs)? That would make testing a lot easier.

Nope, not yet. The assumption basically is so far

1. networking with lwIP works
2. interfacing lwIP with sock works as shown by the test applications

=> networking over sock works.

miri64 commented:
Addressed comments

PeterKietzmann commented:
Hmm. I'm not happy with that answer. Now I have to write it by myself, find out module names and stuff, which costs time.

miri64 commented:
Why? There is a test application provided. If you want I can adapt `tests/lwip` in a separate PR... Anything else doesn't make sense, since the app currently uses both `sock_udp` (introduced in #5937) and `sock_ip` (introduced here) and is supposed to only get `sock_tcp` support (introduced in #5938).

miri64 commented:
> Hmm. I'm not happy with that answer. Now I have to write it by myself, find out module names and stuff, which costs time.

Just to give you a heads-up if you really want to do this: it's a little bit more then that. You need to initialize the socks in code, too. Just adding some modules won't cut it.

miri64 commented:
@PeterKietzmann see #6372

PeterKietzmann commented:
Thanks @miri64 that was really helpful. With#6372 I was able to send IP (as well as UDP) between native nodes and also Atmel boards. Guess I have to scan the code again before a merge.

PeterKietzmann commented:
Ok I think I'd like to merge this soon. But Jenikns states some errors. We can also have a look at Murdock but for that you should squash your fixup-commits.

miri64 commented:
Oh that's just lwIP not liking 8-bit and 16-bit platforms. Found that out when working on #6372.

PeterKietzmann commented:
I was able to `IPV6=1 BOARD=z1 make clean all flash term` and got  `ALL TESTS SUCCESSFUL`

miri64 commented:
which application?

miri64 commented:
(I also get errors, due to some libc functions not defined on `z1`. Are you using some other toolchain/libc than the default Ubuntu one?)

miri64 commented:
```
$ msp430-gcc --version
msp430-gcc (GCC) 4.6.3 20120301 (mspgcc LTS 20120406 unpatched)
Copyright (C) 2011 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

PeterKietzmann commented:
Are there other applications than tests/lwip_sock_ip? My msp430 gcc is
`msp430-gcc (GCC) 4.6.3 20120301`

miri64 commented:
> Are there other applications than tests/lwip_sock_ip? My msp430 gcc is
msp430-gcc (GCC) 4.6.3 20120301

Not in this PR, but might have been that you were referring to #6372. Curious on `z1` I get error messages about missing `fflush()`. This is why I blacklisted it for now (and immediately squashed)

miri64 commented:
Oops, something went wrong.. Fixed now.

miri64 commented:
Removed useless assignment as [reported by Jenkins](https://riot-ci.inet.haw-hamburg.de/job/RIOT/job/PR-5936/5/artifact/error_static-tests.log) (the variable wasn't used after that point, so it was valid) and squashed immediately:

```diff
diff --git a/pkg/lwip/contrib/sock/lwip_sock.c b/pkg/lwip/contrib/sock/lwip_sock.c
index 4dc9bd5..8ae799b 100644
--- a/pkg/lwip/contrib/sock/lwip_sock.c
+++ b/pkg/lwip/contrib/sock/lwip_sock.c
@@ -488,7 +488,6 @@ ssize_t lwip_sock_send(struct netconn **conn, const void *data, size_t len,
             return -EINVAL;
         }
         tmp = *conn;
-        type = (*conn)->type;
     }
     else {
         netbuf_delete(buf);
```

miri64 commented:
Added more boards to `BOARD_INSUFFICIENT_MEMORY` that were reported by Murdock and squashed immediately.

miri64 commented:
Murdock and Jenkins are both happy :-)

miri64 commented:
Sure thing :-).

benpicco commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -0,0 +1,526 @@
+    if (conn == NULL) {
+        return -EADDRNOTAVAIL;
+    }
+#if LWIP_SO_RCVTIMEO
```
What's the reason to make this optional?
It's kind of surprising that LWIP will by default ignore the timeout in e.g. `sock_udp_recv()`
